fedimint-api/src/db/mod.rs

```diff
@@ -880,3 +999,168 @@ mod tests {
         assert_eq!(returned_keys, expected_keys);
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use futures::Future;
+    use tokio::sync::oneshot;
+
+    use super::mem_impl::MemDatabase;
+    use super::*;
+
+    async fn future_returns_shortly<F: Future>(sub: F) -> Option<F::Output> {
+        tokio::time::timeout(Duration::from_millis(10), sub)
+            .await
+            .ok()
+    }
+
+    async fn waiter(db: &Database, key: TestKey) -> tokio::task::JoinHandle<Option<TestVal>> {
+        let db = db.clone();
+        let (tx, rx) = oneshot::channel::<()>();
+        let join_handle = tokio::spawn(async move {
+            let sub = db.wait_key(&key);
+            tx.send(()).unwrap();
+            sub.await.expect("DB error")
+        });
+        rx.await.unwrap();
+        join_handle
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_before_transaction() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new());
+        db.register_module_decoders(ModuleDecoderRegistry::default());
+
+        let key_task = waiter(&db, TestKey(1)).await;
+
+        let mut tx = db.begin_transaction().await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.notify_key(&key).await;
+        tx.commit_tx().await.expect("DB error");
+
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(Some(TestVal(2))),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_before_insert() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new());
+        db.register_module_decoders(ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        let key_task = waiter(&db, TestKey(1)).await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.notify_key(&key).await;
+        tx.commit_tx().await.expect("DB error");
+
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(Some(TestVal(2))),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_after_insert() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new());
+        db.register_module_decoders(ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.notify_key(&key).await;
```

Comment:
Should notify_key be allowed to be called explicitly on the transaction? This test is a tad weird to me, since it is calling notify_key after the insert, but before it is committed to the database. Technically, the record has not been persisted yet and there could be errors between here and commit that would cause this record to not be written.

What is the semantics of notify_key? My initial impression was that it should be called at commit time

Reply:
should notify_key right after you insert the key. It will get queued to be sent at commit time.

Reply:
Thank you, I see it now